bftengine/include/bftengine/CryptoManager.hpp

```diff
@@ -30,73 +33,134 @@ class CryptoManager : public IKeyExchanger, public IMultiSigKeyGenerator {
     static CryptoManager cm_(cryptoSys);
     return cm_;
   }
-
-  std::shared_ptr<IThresholdSigner> thresholdSignerForSlowPathCommit(const std::int64_t& seqnum) const {
-    return thresholdSigner_;
+  std::shared_ptr<IThresholdSigner> thresholdSignerForSlowPathCommit(const SeqNum sn) const {
+    return get(sn)->thresholdSigner_;
   }
-  std::shared_ptr<IThresholdVerifier> thresholdVerifierForSlowPathCommit(const std::int64_t& seqnum) const {
-    return thresholdVerifierForSlowPathCommit_;
+  std::shared_ptr<IThresholdVerifier> thresholdVerifierForSlowPathCommit(const SeqNum sn) const {
+    return get(sn)->thresholdVerifierForSlowPathCommit_;
   }
-
-  std::shared_ptr<IThresholdSigner> thresholdSignerForCommit(const std::int64_t& seqnum) const {
-    return thresholdSigner_;
+  std::shared_ptr<IThresholdSigner> thresholdSignerForCommit(const SeqNum sn) const {
+    return get(sn)->thresholdSigner_;
   }
-  std::shared_ptr<IThresholdVerifier> thresholdVerifierForCommit(const std::int64_t& seqnum) const {
-    return thresholdVerifierForCommit_;
+  std::shared_ptr<IThresholdVerifier> thresholdVerifierForCommit(const SeqNum sn) const {
+    return get(sn)->thresholdVerifierForCommit_;
   }
-
-  std::shared_ptr<IThresholdSigner> thresholdSignerForOptimisticCommit(const std::int64_t& seqnum) const {
-    return thresholdSigner_;
+  std::shared_ptr<IThresholdSigner> thresholdSignerForOptimisticCommit(const SeqNum sn) const {
+    return get(sn)->thresholdSigner_;
   }
-  std::shared_ptr<IThresholdVerifier> thresholdVerifierForOptimisticCommit(const std::int64_t& seqnum) const {
-    return thresholdVerifierForOptimisticCommit_;
+  std::shared_ptr<IThresholdVerifier> thresholdVerifierForOptimisticCommit(const SeqNum sn) const {
+    return get(sn)->thresholdVerifierForOptimisticCommit_;
   }
-
-  // IMultiSigKeyGenerator method
+  // IMultiSigKeyGenerator methodsKEY_EX_LOG
   std::pair<std::string, std::string> generateMultisigKeyPair() override {
-    return multiSigCryptoSystem_->generateNewKeyPair();
+    // use default Cryptosystem for key generation
+    LOG_INFO(logger(), "Generating new multisig key pair");
+    return cryptoSystems_.at(0)->sys_->generateNewKeyPair();
   }
-
   // IKeyExchanger methods
-  void onPrivateKeyExchange(const std::string& secretKey, const std::string& verificationKey) override {
-    multiSigCryptoSystem_->updateKeys(secretKey, verificationKey);
-    init();
+  // onPrivateKeyExchange and onPublicKeyExchange callbacks for a given checkpoint may be called in a different order.
+  // Therefore the first called will create a CryptoSys
+  void onPrivateKeyExchange(const std::string& secretKey,
+                            const std::string& verificationKey,
+                            const SeqNum& sn) override {
+    LOG_INFO(logger(), KVLOG(sn, verificationKey));
+    auto sys = create(sn);
+    sys->sys_->updateKeys(secretKey, verificationKey);
+    sys->init();
```

Comment:
what is the meaning of ```sys```, with ```auto``` it's hard to understand what is this variable.
(also will eliminate the sys->sys_)

Reply:
agree